strawberryfields/engine.py

```diff
@@ -218,43 +218,36 @@ def _broadcast_nones(val, shots):
-                    if not p.can_follow(prev):
-                        raise RuntimeError("Register mismatch: program {}, '{}'.".format(len(self.run_progs), p.name))
-
-                    # Copy the latest measured values in the RegRefs of p.
-                    # We cannot copy from prev directly because it could be used in more than one engine.
-                    for k, v in enumerate(self.samples):
-                        p.reg_refs[k].val = v
-
-                # if the program hasn't been compiled for this backend, do it now
-                if p.backend != self.backend_name:
-                    p = p.compile(self.backend_name, **compile_options) # TODO: shots might be relevant for compilation?
-                p.lock()
-
-                kwargs["shots"] = shots
-                # Note: by putting ``shots`` into keyword arguments, it allows for the
-                # signatures of methods in Operations to remain cleaner, since only
-                # Measurements need to know about shots
-
-                self._run_program(p, **kwargs)
-                self.run_progs.append(p)
-                # store the latest measurement results
-                shots = kwargs.get("shots", 1)
-                self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
-                prev = p
-        except Exception as e:
-            raise e
-        else:
-            # program execution was successful
-            pass
+        kwargs["shots"] = shots
+        # NOTE: by putting ``shots`` into keyword arguments, it allows for the
+        # signatures of methods in Operations to remain cleaner, since only
+        # Measurements need to know about shots
+
+        prev = self.run_progs[-1] if self.run_progs else None  # previous program segment
+        for p in program:
+            if prev is None:
+                # initialize the backend
+                self._init_backend(p.init_num_subsystems)
+            else:
+                # there was a previous program segment
+                if not p.can_follow(prev):
+                    raise RuntimeError("Register mismatch: program {}, '{}'.".format(len(self.run_progs), p.name))
+
+                # Copy the latest measured values in the RegRefs of p.
+                # We cannot copy from prev directly because it could be used in more than one engine.
+                for k, v in enumerate(self.samples):
+                    p.reg_refs[k].val = v
+
+            # if the program hasn't been compiled for this backend, do it now
+            if p.target != self.backend_name:
+                p = p.compile(self.backend_name, **compile_options) # TODO: shots might be relevant for compilation?
+            p.lock()
+
+            self._run_program(p, **kwargs)
+            self.run_progs.append(p)
+            # store the latest measurement results
+            self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
```

Comment:
Should we potentially rename `broadcast_nones` to `_normalize` or `_normalize_shape` or `_fix_shape`? I would also suggest rewriting this for readability/clarity as:

```
            self.samples = map(_normalize_sample, sorted(p.reg_refs))
```

Where `_normalize_sample` takes `p.reg_ref`. Below, we would need to do ```return Result(list(self.samples))``` or similar.

Reply:
I think `_broadcast_nones` won't be needed anymore once #107 is done.

Reply:
This piece of code will be made obsolete by the refactor in #111

Reply:
Off topic, but worth noting that most Python style guides recommend list comprehensions over map/filter/reduce, as being more clear and explicit (and more 'Pythonic', although that is debatable).

There is also a slight functionality difference in that `map` returns an iterator, not a list, so they cannot always be directly swapped out - iterators are stateful, and that they mutate as you traverse them.

Reply:
@josh146 agreed in most cases, however in my opinion when a list comprehension looks like `[f(x) for x in s]`, it is a little more readable to write `map(f, s)`. In my original example as I'd mentioned, you'd have to return ```list(self.samples)``` if you chose to use map. In any case -- moot point I suppose, since there was an extra step of sorting the values, not the keys, that I have omitted.

Reply:
@zeid, yeah, I agree with your use case for map.